src/server/session.rs

```diff
@@ -92,7 +92,231 @@ where
+    /// Processing for the data connection. This will spawn a new async task with the actual processing.
+    ///
+    /// socket: the data socket we'll be working with
+    /// tls: tells if this should be a TLS connection
+    /// tx: channel to send the result of our operation to the control process
+    pub(super) fn spawn_data_processing(&mut self, socket: tokio02::net::TcpStream, tx: Sender<InternalMsg>) {
+        let mut data_cmd_rx = self.data_cmd_rx.take().unwrap().fuse();
+        let mut data_abort_rx = self.data_abort_rx.take().unwrap().fuse();
+        let tls = self.data_tls;
+        let mut command_executor = DataCommandExecutor {
+            user: self.user.clone(),
+            socket,
+            tls,
+            tx,
+            cmd: None,
+            storage: Arc::clone(&self.storage),
+            cwd: self.cwd.clone(),
+            start_pos: self.start_pos,
+            identity_file: if tls { Some(self.certs_file.clone().unwrap()) } else { None },
+            identity_password: if tls { Some(self.certs_password.clone().unwrap()) } else { None },
+        };
+
+        tokio02::spawn(async move {
+            #[allow(unused_assignments)]
+            let mut incoming = None;
+            let mut timeout_delay = tokio02::time::delay_for(std::time::Duration::from_secs(5 * 60));
+            // TODO: Use configured timeout
+            tokio02::select! {
+                Some(command) = data_cmd_rx.next() => {
+                    incoming = Some(DataCommand::ExternalCommand(command));
+                },
+                Some(_) = data_abort_rx.next() => {
+                    incoming = Some(DataCommand::Abort);
+                },
+                _ = &mut timeout_delay => {
+                    info!("Connection timed out");
+                    return;
+                }
+            };
+
+            match incoming {
+                Some(DataCommand::Abort) => {
+                    info!("Abort received");
+                }
+                Some(DataCommand::ExternalCommand(command)) => {
+                    info!("Data command received");
+                    command_executor.cmd = Some(command);
+                    command_executor.execute().await;
+                }
+                None => {
+                    // This probably happened because the control channel was closed before we got here
+                    warn!("Nothing received");
+                }
+            }
+        });
+    }
+}
+
+impl<S, U: Send + Sync> Drop for Session<S, U>
+where
+    S: storage::StorageBackend<U>,
+    S::File: crate::storage::AsAsyncReads + Send,
+    S::Metadata: storage::Metadata,
+{
+    fn drop(&mut self) {
+        if self.with_metrics {
+            // Decrease the sessions metrics gauge when the session goes out of scope.
+            metrics::dec_session();
+        }
+    }
+}
+
+struct DataCommandExecutor<S, U: Send + Sync>
```

Comment:
This struct's definition and its implementation  can now be moved to a `datachannel` module or sometthing of the like.